### PR TITLE
Azure integration

### PR DESCRIPTION
**Pull request by aluminiumgeek**
Adds provider modules needed for integration with AKS.

Including:
- cluster initialization changes
- updated docs
- azuresql
- sql connection string updates to be compatible with azuresql format

### CODE REVIEW

File: ckan_cloud_operator/deis_ckan/envvars.py [hunk 1]
```diff
@@ -57,10 +57,10 @@ def _update(self):
             postgres_host, postgres_port = db_manager.get_internal_proxy_host_port()
             logs.info(f'Connecting to DB proxy: {postgres_host}:{postgres_port}')
         envvars.update(
-            CKAN_SQLALCHEMY_URL=f"postgresql://{db_name}:{db_password}@{postgres_host}:{postgres_port}/{db_name}",
-            CKAN___BEAKER__SESSION__URL=f"postgresql://{db_name}:{db_password}@{postgres_host}:{postgres_port}/{db_name}",
-            CKAN__DATASTORE__READ_URL=f"postgresql://{datastore_ro_user}:{datastore_ro_password}@{postgres_host}:{postgres_port}/{datastore_name}",
-            CKAN__DATASTORE__WRITE_URL=f"postgresql://{datastore_name}:{datastore_password}@{postgres_host}:{postgres_port}/{datastore_name}",
+            CKAN_SQLALCHEMY_URL=f"postgresql://{db_name}@{postgres_host}:{db_password}@{postgres_host}:{postgres_port}/{db_name}",
```
Comment (akariv) [inline ckan_cloud_operator/deis_ckan/envvars.py]: Why this change? it makes the urls invalid (and how are they related to azure?)

File: ckan_cloud_operator/deis_ckan/instance.py [hunk 2]
```diff
@@ -312,7 +312,7 @@ def update(self, wait_ready=False, skip_solr=False, skip_deployment=False):
         except Exception:
             logs.warning('Setting datastore permissions failed, continuing anyway')
         # Create/Update uptime monitoring after everything else is ready
-        DeisCkanInstanceUptime(self).update(envvars.site_url)
+        #DeisCkanInstanceUptime(self).update(envvars.site_url)
```
Comment (akariv) [inline ckan_cloud_operator/deis_ckan/instance.py]: why?

File: ckan_cloud_operator/labels/manager.py [hunk 3]
```diff
@@ -11,6 +11,7 @@ def initialize(log_kwargs=None):
 
 def get_label_prefix(short=False):
     """Returns a global label prefix which should be used to namespace operator objects"""
+    return 'ckan-cloud'
```
Comment (akariv) [inline ckan_cloud_operator/labels/manager.py]: looks like some debug code leftover (?)

File: ckan_cloud_operator/providers/ckan/db/migration.py [hunk 4]
```diff
@@ -438,7 +438,8 @@ def _create_base_dbs_and_roles(migration_name, db_name, datastore_name, recreate
             postgres_driver.create_role_if_not_exists(admin_conn, datastore_ro_name, datastore_ro_password)
             yield {'step': 'created-datastore-ro-role', 'msg': f'Created Datastore read-only user: {datastore_ro_name} ({db_prefix})'}
     yield {'step': 'created-base-dbs-and-roles', f'msg': f'Created base dbs and roles: {db_name}, {datastore_name}, {datastore_ro_name} ({db_prefix})'}
-    yield from _update_db_proxy(db_name, datastore_name, datastore_ro_name, db_password, datastore_password, datastore_ro_password, db_prefix)
+    if cluster_manager.get_provider_id() != 'azure':
```
Comment (akariv) [inline ckan_cloud_operator/providers/ckan/db/migration.py]: why?

File: ckan_cloud_operator/providers/cluster/azure/manager.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
+#### standard provider code ####
+
+import subprocess
```
Comment (akariv) [inline ckan_cloud_operator/providers/cluster/azure/manager.py]: move to imports section below

File: ckan_cloud_operator/providers/cluster/azure/manager.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        }
+    })
+    return {'persistentVolumeClaim': {'claimName': disk_id}}
+
+
+def _generate_password(l):
+    return binascii.hexlify(os.urandom(l)).decode()
+
+
+def check_output(cmd):
```
Comment (akariv) [inline ckan_cloud_operator/providers/cluster/azure/manager.py]: better rename to something more descriptive, e.g. `az_check_output`

File: ckan_cloud_operator/providers/db/azuresql/cli.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+import click
+
+from ckan_cloud_operator import logs
+from . import manager
+
+
+@click.group('azuresql')
+def rds():
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/azuresql/cli.py]: is rds an azure term?

File: ckan_cloud_operator/providers/db/azuresql/manager.py [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+import datetime
+import os
+import yaml
+import subprocess
+import json
+
+from ckan_cloud_operator import logs
+from ckan_cloud_operator.providers.cluster import manager as cluster_manager
+from ckan_cloud_operator.drivers.gcloud import driver as gcloud_driver
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/azuresql/manager.py]: why this import?

File: ckan_cloud_operator/providers/db/azuresql/manager.py [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+              " - SQL instance name\n"
+              "\n")
+        _config_interactive_set(default_values, **_get_config_credentials_kwargs(db_prefix))
+    config = _config_get(**_get_config_credentials_kwargs(db_prefix))
+    for key, default_value in default_values.items():
+        if key not in config and default_value:
+            _config_set(key, default_value, **_get_config_credentials_kwargs(db_prefix))
+        elif not config.get(key):
+            raise Exception(f'missing key: {key}')
+    #from ckan_cloud_operator.providers.db.proxy.azuresql import manager as proxy_azuresql_manager
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/azuresql/manager.py]: remove commented out code

File: ckan_cloud_operator/providers/storage/azure/manager.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+from ckan_cloud_operator import logs, kubectl
+from ckan_cloud_operator.config import manager as config_manager
+from ckan_cloud_operator.providers.cluster.azure.manager import _config_get as cluster_config_get, check_output as az_check_output
+
+from ..constants import CONFIG_NAME
+from .constants import PROVIDER_ID
+
+
+def initialize(*args, **kwargs):
```
Comment (akariv) [inline ckan_cloud_operator/providers/storage/azure/manager.py]: please provide explicit parameters to this function

File: ckan_cloud_operator/deis_ckan/envvars.py [hunk 11]
```diff
@@ -57,10 +57,10 @@ def _update(self):
...
+            CKAN___BEAKER__SESSION__URL=f"postgresql://{db_name}@{postgres_host}:{db_password}@{postgres_host}:{postgres_port}/{db_name}",
+            CKAN__DATASTORE__READ_URL=f"postgresql://{datastore_ro_user}@{postgres_host}:{datastore_ro_password}@{postgres_host}:{postgres_port}/{datastore_name}",
+            CKAN__DATASTORE__WRITE_URL=f"postgresql://{datastore_name}@{postgres_host}:{datastore_password}@{postgres_host}:{postgres_port}/{datastore_name}",
```
Comment (zelima) [inline ckan_cloud_operator/deis_ckan/envvars.py]: Could you please explain this? Why do we need `postgres_host` before username as well?

Comment (zelima) [reply ckan_cloud_operator/deis_ckan/envvars.py]: Also, I remember I had a problem with connecting to azure database with URIs like this as username include `@` in it. Eg `psql postgresql://irakli.mchedlishvli@azure.com:passowrd@host:5000` was erroring cause it's getting confused with two `@`s in URL. Have we tested CKAN is fine with that?

File: ckan_cloud_operator/deis_ckan/instance.py (hunk 2 above)
Comment (zelima) [inline ckan_cloud_operator/deis_ckan/instance.py]: Should we just delete it if not needed?

File: ckan_cloud_operator/deis_ckan/instance.py [hunk 12]
```diff
@@ -338,7 +338,7 @@ def delete(self, force=False, wait_deleted=False):
                 'namespace': lambda: DeisCkanInstanceNamespace(self).delete(),
                 'envvars-secret': lambda: kubectl.check_call(f'delete --ignore-not-found secret/{self.id}-envvars'),
                 'routes': lambda: routers_manager.delete_routes(deis_instance_id=self.id),
-                'uptime-monitoring': lambda: DeisCkanInstanceUptime(self).delete(self.id)
+                #'uptime-monitoring': lambda: DeisCkanInstanceUptime(self).delete(self.id)
```
Comment (zelima) [inline ckan_cloud_operator/deis_ckan/instance.py]: Same here, do we actually need this comment?

File: ckan_cloud_operator/labels/manager.py (hunk 3 above)
Comment (zelima) [inline ckan_cloud_operator/labels/manager.py]: I assume this is hardcoded for development purposes... (Or we don't need another `return` coming after)

File: ckan_cloud_operator/providers/ckan/db/migration.py [hunk 13]
```diff
@@ -438,7 +438,8 @@ def _create_base_dbs_and_roles(migration_name, db_name, datastore_name, recreate
...
+        yield from _update_db_proxy(db_name, datastore_name, datastore_ro_name, db_password, datastore_password, datastore_ro_password, db_prefix)
```
Comment (zelima) [inline ckan_cloud_operator/providers/ckan/db/migration.py]: Why not doing this for azure?

File: ckan_cloud_operator/providers/cluster/azure/manager.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+
+def initialize(interactive=False):
+    _set_provider()
+    if interactive:
+        print('\nEnter the Resource Group name\n')
+        _config_interactive_set({'azure-rg': None}, is_secret=False)
+        print('\nEnter the location of the Kubernets cluster is hosted on [westus2]\n')
+        _config_interactive_set({'azure-default-location': None}, is_secret=False)
+        print('\nEnter the name of your cluster\n')
+        _config_interactive_set({'azure-cluster-name': None}, is_secret=False)
```
Comment (zelima) [inline ckan_cloud_operator/providers/cluster/azure/manager.py:39]: What if not interactive? Will it just skip and do nothing? Should not we at least say smth like "please set --interactive flag" or similar?

File: ckan_cloud_operator/providers/db/azuresql/manager.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    #proxy_azuresql_manager.initialize(db_prefix=db_prefix)
```
Comment (zelima) [inline ckan_cloud_operator/providers/db/azuresql/manager.py]: Redundant comments?

File: docs/PRODUCTION-AZURE-CLUSTER.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+```
+kubectl create namespace ckan-cloud
+kubectl -n ckan-cloud create secret generic ckan-cloud-provider-cluster-gcloud
+kubectl -n ckan-cloud create configmap operator-conf --from-literal=ckan-cloud-operator-image=viderum/ckan-cloud-operator:latest --from-literal=label-prefix=ckan-cloud
+```
+
+After that create `uptime-statuscake-api` secrets with keys "user", "key", "group" populated from StatusCake account:
+```
+kubectl -n ckan-cloud create secret generic uptime-statuscake-api --from-literal=user=<user> --from-literal=key=<key> --from-literal=group=<group>
+```
```
Comment (zelima) [inline docs/PRODUCTION-AZURE-CLUSTER.md]: Why not this is the part of `ckan-cloud-operator cluster initialize --interactive`?

Comment (zelima) [reply docs/PRODUCTION-AZURE-CLUSTER.md]: Also, think statuscake stuff should be optional... What if I don't want to monitor uptime at all? Or want a different providers?

File: docs/PRODUCTION-AZURE-CLUSTER.md [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+Then run interactive initialization of the currently connected cluster:
+```
+ckan-cloud-operator cluster initialize --interactive
+```
+
+While interactive initialization:
+- Set `enable-deis-ckan: y`
+- If environment is production, set `env-id` to `p` on "routers" step.
+- On "solr" step of interactive initialization choose `self-hosted: y`
```
Comment (zelima) [inline docs/PRODUCTION-AZURE-CLUSTER.md]: Just wondering, will this work if I set to `n` instead? Or I have to agree if I need to continue?

File: docs/PRODUCTION-AZURE-CLUSTER.md [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+- On "ckan" step when asked for docker server/username/password, enter your Gitlab credentials, password should be your Gitlab access token.
+
+
+Give the service account permission to change cluster roles:
+```
+kubectl create clusterrolebinding default-sa-binding --clusterrole=cluster-admin --user=<service account email>
+```
```
Comment (zelima) [inline docs/PRODUCTION-AZURE-CLUSTER.md]: Again kubectl command. Think we should wrap it with `ckan-cloud-opeartor`, no?

**Review by zelima — COMMENTED** [review]
Great work! Mostly have some questions and suggestions, would be great if you could answer

File: docs/PRODUCTION-AZURE-CLUSTER.md [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+
+Log-in to your Rancher deployment on the management server.
+
+Add cluster > Import existing cluster > Follow instructions in the UI
+
+Click on the cluster and then on kubeconfig file.
+
+Download the file locally.
+
+## Create multi-user storage class
```
Comment (akariv) [inline docs/PRODUCTION-AZURE-CLUSTER.md]: This also needs to be done in CCO initialisation.
See the AWS code (once you've rebased).

**Review by akariv — CHANGES_REQUESTED** [review]

File: ckan_cloud_operator/labels/manager.py (hunk 3 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/labels/manager.py]: right

File: ckan_cloud_operator/providers/ckan/db/migration.py (hunk 4 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/ckan/db/migration.py]: DB proxy module for azure wasn't ready. I pushed it.

File: ckan_cloud_operator/providers/db/azuresql/manager.py (hunk 9 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/azuresql/manager.py]: Uncommented it, as azuresql proxy module is ready

File: ckan_cloud_operator/deis_ckan/envvars.py (hunk 1 above)
Comment (akariv) [reply ckan_cloud_operator/deis_ckan/envvars.py]: @aluminiumgeek

File: ckan_cloud_operator/providers/cluster/gcloud/manager.py [hunk 20]
```diff
@@ -42,6 +42,8 @@ def initialize(interactive=False):
         _config_interactive_set({'cluster-name': None})
         print('\nEnter the google project ID\n')
         _config_interactive_set({'project-id': None})
+    else:
```
Comment (akariv) [inline ckan_cloud_operator/providers/cluster/gcloud/manager.py]: why this log message? you're not really skipping anything really...

File: ckan_cloud_operator/providers/db/proxy/azuresql/cli.py [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+import traceback
+
+from ckan_cloud_operator import logs
+
+from . import manager
+
+
+@click.group()
+def azuresql():
+    """Manage the RDS centralized db proxy"""
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/proxy/azuresql/cli.py]: another copy paste :)

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+def _apply_deployment(db_prefix=None):
+    db_host = config_manager.get(secret_name='ckan-cloud-provider-db-azuresql-credentials')['azuresql-host']
+    kubectl.apply(kubectl.get_deployment(
+        _get_resource_name(suffix=db_prefix),
+        _get_resource_labels(for_deployment=True, suffix=db_prefix or ''),
+        {
+            'replicas': 1,
+            'revisionHistoryLimit': 10,
+            'strategy': {'type': 'RollingUpdate', },
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/proxy/azuresql/manager.py:105]: this won't work for k8s 1.16

File: ckan_cloud_operator/providers/storage/azure/constants.py [hunk 23]
```diff
@@ -0,0 +1 @@
+PROVIDER_ID = 'azure'
```
Comment (akariv) [inline ckan_cloud_operator/providers/storage/azure/constants.py]: use a more accurate provider id ('azure' is already in use)

**Review by akariv — CHANGES_REQUESTED** [review]
Some unanswered questions still...
+ some modifications in the deis code which shouldn't be

File: tests/routers/test_traefik.py [hunk 24]
```diff
@@ -36,7 +36,7 @@ def test_create_with_cloudflare_without_credentials(self, update_flag, get_provi
                 'name': 'datapusher',
             },
             'spec': {
-                'default-root-domain': 'default',
```
Comment (akariv) [inline tests/routers/test_traefik.py]: If the purpose of this test is to check that we assert on no credentials, then this fix won't be testing that.

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py [hunk 25]
```diff
@@ -102,6 +102,11 @@ def _apply_deployment(db_prefix=None):
             'replicas': 1,
             'revisionHistoryLimit': 10,
             'strategy': {'type': 'RollingUpdate', },
+            'selector': {
+                'matchLabels': {
+                    'app': 'provider-db-proxy-azuresql'
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/proxy/azuresql/manager.py]: we have a function to generate the labels - why not use it?
(i.e. `_get_resource_labels(for_deployment=True, suffix=db_prefix or ''),`)

**Review by akariv — CHANGES_REQUESTED** [review]

**Comment by OriHoch** [comment]
:+1:

File: ckan_cloud_operator/deis_ckan/envvars.py (hunk 1 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/deis_ckan/envvars.py]: Azure SQL instances support usernames only is this format ("username@host"). This format in URI works with patched version of CKAN for Gates.

I'll revert this change anyway, as we currently use helm based CKAN instances instead of deis.

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py (hunk 22 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/proxy/azuresql/manager.py:105]: I don't understand what will not work. Could you explain, please?

File: ckan_cloud_operator/providers/storage/azure/constants.py (hunk 23 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/storage/azure/constants.py]: I'll rename it to `abs` (for Azure Blob Storage)

File: docs/PRODUCTION-AZURE-CLUSTER.md (hunk 16 above)
Comment (aluminiumgeek) [reply docs/PRODUCTION-AZURE-CLUSTER.md]: I removed info about it, because it's related only to deis instances.

File: docs/PRODUCTION-AZURE-CLUSTER.md (hunk 17 above)
Comment (aluminiumgeek) [reply docs/PRODUCTION-AZURE-CLUSTER.md]: It will ask for http endpoint of an external solr instance.

File: docs/PRODUCTION-AZURE-CLUSTER.md (hunk 18 above)
Comment (aluminiumgeek) [reply docs/PRODUCTION-AZURE-CLUSTER.md]: It's related only to GKE clusters, removed.

File: tests/routers/test_traefik.py (hunk 24 above)
Comment (aluminiumgeek) [reply tests/routers/test_traefik.py]: Got it

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py (hunk 22 above)
Comment (akariv) [reply ckan_cloud_operator/providers/db/proxy/azuresql/manager.py:105]: afaik on k8s 1.16 it won't work without a `selector` entry.

File: ckan_cloud_operator/providers/storage/azure/constants.py (hunk 23 above)
Comment (akariv) [reply ckan_cloud_operator/providers/storage/azure/constants.py]: please resolve only when done...

**Comment by akariv** [comment]
@aluminiumgeek 
please add a comment 'done' or 'fixed' when a CR issue was fixed.
it's up to the reviewer to resolve the issue or not.

File: ckan_cloud_operator/providers/solr/solrcloud/manager.py [hunk 26]
```diff
@@ -94,13 +94,11 @@ def initialize(interactive=False, dry_run=False):
...
 
-    # TODO - need to check the pod names and ensure these are solrcloud ones
-    # TODO - actual number is _+ 1_ and not _+ 2_
-    expected_running = len(sc_host_names) + len(zk_host_names) + 2
+    expected_running = len(sc_host_names) + len(zk_host_names)
     while True:
         pods = kubectl.get('pods')
         running = len([x for x in pods['items']
-                       if x['status']['phase'] == 'Running'])
+                       if x['status']['phase'] == 'Running' and ('solrcloud-zk' in x['metadata']['name'] or 'solrcloud-sc' in x['metadata']['name'])])
```
Comment (akariv) [inline ckan_cloud_operator/providers/solr/solrcloud/manager.py]: see my other comment on checking the ['labels']['app'] instead.

File: ckan_cloud_operator/providers/db/proxy/azuresql/constants.py [hunk 27]
```diff
@@ -0,0 +1 @@
+PROVIDER_ID='azuresql'
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/proxy/azuresql/constants.py]: the proxy can't be the same provider id as the main db provider

File: ckan_cloud_operator/providers/cluster/azure/manager.py (hunk 14 above)
Comment (akariv) [reply ckan_cloud_operator/providers/cluster/azure/manager.py:39]: providers_manager.config_interactive_set has an 'interactive' parameter, better use it instead - that is, remove the if and pass interactive to every call of `_config_interactive_set`

File: ckan_cloud_operator/providers/cluster/aws/manager.py [hunk 28]
```diff
@@ -41,6 +41,8 @@ def initialize(interactive=False):
         _config_interactive_set({'aws-default-region': None}, is_secret=True)
         print('\nEnter the name of your Amazon EKS cluster\n')
         _config_interactive_set({'eks-cluster-name': None}, is_secret=True)
+    else:
```
Comment (akariv) [inline ckan_cloud_operator/providers/cluster/aws/manager.py]: see comment below... about this block

File: ckan_cloud_operator/providers/db/proxy/azuresql/constants.py (hunk 27 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/proxy/azuresql/constants.py]: `gcloudsql` and `rds` db providers has the same provider id as their proxies

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py (hunk 22 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/proxy/azuresql/manager.py:105]: It has a selector entry now.

File: ckan_cloud_operator/providers/cluster/aws/manager.py (hunk 28 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/cluster/aws/manager.py]: done

File: ckan_cloud_operator/providers/solr/solrcloud/manager.py (hunk 26 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/solr/solrcloud/manager.py]: done

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py (hunk 25 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/proxy/azuresql/manager.py]: done

File: ckan_cloud_operator/providers/storage/azure/constants.py (hunk 23 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/storage/azure/constants.py]: done

File: ckan_cloud_operator/providers/solr/solrcloud/manager.py [hunk 29]
```diff
@@ -97,8 +97,10 @@ def initialize(interactive=False, dry_run=False):
     expected_running = len(sc_host_names) + len(zk_host_names)
     while True:
         pods = kubectl.get('pods')
+        sc_app_name = _get_resource_labels(for_deployment=True, suffix='sc')['app']
+        zk_app_name = _get_resource_labels(for_deployment=True, suffix='zk')['app']
         running = len([x for x in pods['items']
-                       if x['status']['phase'] == 'Running' and ('solrcloud-zk' in x['metadata']['name'] or 'solrcloud-sc' in x['metadata']['name'])])
+                       if x['status']['phase'] == 'Running' and (zk_app_name in x['metadata']['name'] or sc_app_name in x['metadata']['name'])])
```
Comment (akariv) [inline ckan_cloud_operator/providers/solr/solrcloud/manager.py]: please use x['metadata']['labels']['app']

File: ckan_cloud_operator/providers/db/proxy/azuresql/constants.py (hunk 27 above)
Comment (akariv) [reply ckan_cloud_operator/providers/db/proxy/azuresql/constants.py]: I guess that's also a potential bug, as they might overwrite one another's configuration... unless there's a hidden reason for this.

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py (hunk 25 above)
Comment (akariv) [reply ckan_cloud_operator/providers/db/proxy/azuresql/manager.py]: where?

File: ckan_cloud_operator/logs.py [hunk 30]
```diff
@@ -74,7 +74,7 @@ def log_subprocess_output(stdout, stderr):
 def subprocess_run(command, input=None):
     completed = subprocess.run(
         command, input=input, 
-        shell=True, check=True, capture_output=True
+        shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
```
Comment (akariv) [inline ckan_cloud_operator/logs.py]: why?

File: ckan_cloud_operator/providers/ckan/db/migration.py [hunk 31]
```diff
@@ -81,12 +81,12 @@ def delete(name, delete_dbs=False):
     migration = crds_manager.get(CRD_SINGULAR, name=name, required=False) or {}
```
Comment (akariv) [inline ckan_cloud_operator/providers/ckan/db/migration.py]: why is this file being modified?

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py [hunk 32]
```diff
@@ -0,0 +1,156 @@
...
+    kubectl.apply(kubectl.get_deployment(
+        _get_resource_name(suffix=db_prefix),
+        _get_resource_labels(for_deployment=True, suffix=db_prefix or ''),
+        {
+            'replicas': 1,
+            'revisionHistoryLimit': 10,
+            'strategy': {'type': 'RollingUpdate', },
+            'selector': {
+                'matchLabels': {
+                    'app': app_name
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/proxy/azuresql/manager.py]: please use the proper method for getting the labels

File: ckan_cloud_operator/logs.py (hunk 30 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/logs.py]: Compatibility with Python 3.6 (`capture_output` introduced in 3.7 and does [the same](https://github.com/python/cpython/blob/3.7/Lib/subprocess.py#L481))

File: docs/PRODUCTION-AZURE-CLUSTER.md [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+Follow the ckan-cloud-operator installation and usage guide in the [README.md](/README.md) to configure ckan-cloud-operator to use with kubeconfig file.
+
+### Initialize the cluster
+Then run interactive initialization of the currently connected cluster:
+```
+ckan-cloud-operator cluster initialize --interactive
+```
+
+While interactive initialization:
+- Set `enable-deis-ckan: y`
```
Comment (akariv) [inline docs/PRODUCTION-AZURE-CLUSTER.md]: remove deis references.

File: ckan_cloud_operator/providers/ckan/db/migration.py (hunk 31 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/ckan/db/migration.py]: Azure SQL requires username in `username@host` format. Connection string with two `@` is not supported in psycopg2, so we pass connection parameters as kwargs.

Comment (akariv) [reply ckan_cloud_operator/providers/ckan/db/migration.py]: The question is what is db_migration being used for in the context of Azure?

File: ckan_cloud_operator/providers/db/proxy/azuresql/constants.py (hunk 27 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/proxy/azuresql/constants.py]: done

File: ckan_cloud_operator/providers/solr/solrcloud/manager.py (hunk 29 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/solr/solrcloud/manager.py]: done

File: ckan_cloud_operator/providers/ckan/db/migration.py (hunk 31 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/ckan/db/migration.py]: It has been used when I set it up with deis instances. It's not used anymore. Reverting it.

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py (hunk 32 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/db/proxy/azuresql/manager.py]: done

File: docs/PRODUCTION-AZURE-CLUSTER.md (hunk 33 above)
Comment (aluminiumgeek) [reply docs/PRODUCTION-AZURE-CLUSTER.md]: done

File: ckan_cloud_operator/providers/solr/solrcloud/manager.py [hunk 34]
```diff
@@ -94,13 +94,13 @@ def initialize(interactive=False, dry_run=False):
...
-    # TODO - actual number is _+ 1_ and not _+ 2_
-    expected_running = len(sc_host_names) + len(zk_host_names) + 2
+    expected_running = len(sc_host_names) + len(zk_host_names)
     while True:
         pods = kubectl.get('pods')
+        sc_app_name = _get_resource_labels(for_deployment=True, suffix='sc')['app']
+        zk_app_name = _get_resource_labels(for_deployment=True, suffix='zk')['app']
         running = len([x for x in pods['items']
-                       if x['status']['phase'] == 'Running'])
+                       if x['status']['phase'] == 'Running' and x['metadata']['labels']['app'] in [zk_app_name, sc_app_name]])
```
Comment (akariv) [inline ckan_cloud_operator/providers/solr/solrcloud/manager.py]: X['metadata']['labels']['app'].startswith('provider-solr-solrcloud')
But replace the string literal with the proper constants

**Comment by akariv** [comment]
@aluminiumgeek 
One comment left, but also please make sure the tests pass.

File: ckan_cloud_operator/providers/solr/solrcloud/manager.py (hunk 34 above)
Comment (aluminiumgeek) [reply ckan_cloud_operator/providers/solr/solrcloud/manager.py]: done

File: docs/DNS-AND-ROUTING.md [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+After that CCO reconfigures Traefik config file to point a domain to the instance, and creates a Lets Encrypt SSL certificate.  
+
+## Supported DNS providers
+- CloudFlare
+- Route 53
+- UltraDNS
+
+Code reference: `ckan_cloud_operator/routers`, `ckan_cloud_operator/providers/cluster/<cloud_provider_name>`
+
+## Adding a new DNS integration
```
Comment (akariv) [inline docs/DNS-AND-ROUTING.md]: Please move this bit to a doc under development/

File: docs/PRODUCTION-AZURE-CLUSTER.md [hunk 36]
```diff
@@ -0,0 +1,53 @@
+# Creating a Production CKAN Cloud cluster using Azure Cloud Platform
```
Comment (akariv) [inline docs/PRODUCTION-AZURE-CLUSTER.md:1]: Please follow the style in the AWS file and connect with the README (both recently modified... :) )

**Review by akariv — CHANGES_REQUESTED** [review]
nearly there...

File: docs/PRODUCTION-AZURE-CLUSTER.md (hunk 36 above)
Comment (aluminiumgeek) [reply docs/PRODUCTION-AZURE-CLUSTER.md:1]: AWS file contains info about terraform which is not ready for Azure yet (in this branch).

I modified the doc adding prerequisites and removing not used data.

File: docs/DNS-AND-ROUTING.md (hunk 35 above)
Comment (aluminiumgeek) [reply docs/DNS-AND-ROUTING.md]: Done

File: ckan_cloud_operator/providers/cluster/manager.py [hunk 37]
```diff
@@ -184,7 +184,7 @@ def create_volume(disk_size_gb, labels, use_existing_disk_name=None, zone=0):
         labels,
         **labels_manager.get_resource_labels(label_suffixes=_get_cluster_volume_label_suffixes())
     )
-    return get_provider().create_volume(disk_size_gb, labels, use_existing_disk_name=use_existing_disk_name, zone=zone)
```
Comment (akariv) [inline ckan_cloud_operator/providers/cluster/manager.py]: why did you remove the zone parameter?

File: ckan_cloud_operator/providers/db/proxy/azuresql/manager.py [hunk 38]
```diff
@@ -0,0 +1,153 @@
...
+    db_host = config_manager.get(secret_name='ckan-cloud-provider-db-azuresql-credentials')['azuresql-host']
+    kubectl.apply(kubectl.get_deployment(
+        _get_resource_name(suffix=db_prefix),
+        _get_resource_labels(for_deployment=True, suffix=db_prefix or ''),
+        {
+            'replicas': 1,
+            'revisionHistoryLimit': 10,
+            'strategy': {'type': 'RollingUpdate', },
+            'selector': {
+                'matchLabels': _get_resource_labels(for_deployment=True, suffix=db_prefix or '')
```
Comment (akariv) [inline ckan_cloud_operator/providers/db/proxy/azuresql/manager.py]: use a `labels` variable instead of calling the same function 3 times

File: ckan_cloud_operator/routers/traefik/deployment.py [hunk 39]
```diff
@@ -89,6 +89,21 @@ def _get_deployment_spec(router_name, router_type, annotations, image=None, http
             'CLOUDFLARE_API_KEY': cloudflare_api_key,
         }, labels=get_labels(router_name, router_type))
         container['envFrom'] = [{'secretRef': {'name': secret_name}}]
+    elif dns_provider == 'azure':
+        logs.info('Traefik deployment: adding SSL support using Azure DNS')
+        container = deployment_spec['template']['spec']['containers'][0]
+        container['ports'].append({'containerPort': 443})
+        azure_credendials = cluster_manager.get_provider().get_azure_credentials()
+        secret_name = f'ckancloudrouter-{router_name}-azure'
+        print(azure_credendials)
```
Comment (akariv) [inline ckan_cloud_operator/routers/traefik/deployment.py]: remove

File: docs/VALUES-FILE-REFERENCE.md [hunk 40]
```diff
@@ -89,6 +91,7 @@
 
 - `useCloudStorage` - use a cloud storage bucket for storing CKAN files
 - `ckanPrimaryColor` - set the newly created ckan theme to be based on this color (hex based RGB, e.g. `#123456`)
+- `cronjobs` - list of CronJob entries to be created on the cluster
```
Comment (akariv) [inline docs/VALUES-FILE-REFERENCE.md]: please add some details on what the value here should look like

File: docs/VALUES-FILE-REFERENCE.md [hunk 41]
```diff
@@ -14,6 +14,8 @@
 - `withSansSSL` - Ignored, will always be considered as True
 - `registerSubdomain` - Ignored, will always be equal to `sub-domain`
 
+- `imagePullSecret` - secret contains credentials for the private docker repo (secret is being created by CCO)
```
Comment (akariv) [inline docs/VALUES-FILE-REFERENCE.md]: please add some details on what the secret should contain

**Review by akariv — CHANGES_REQUESTED** [review]

File: ckan_cloud_operator/providers/cluster/manager.py (hunk 37 above)
Comment (zelima) [reply ckan_cloud_operator/providers/cluster/manager.py]: Don't remember. Probably deleted while debugging and left. Bringing back

**Review by akariv — DISMISSED** [review]

**Review by akariv — APPROVED** [review]
